src/zope/schema/_bootstrapfields.py

```diff
@@ -655,3 +676,176 @@ class Int(Integral):
     """
     _type = integer_types
     _unicode_converters = (int,)
+
+
+VALIDATED_VALUES = threading.local()
```

Comment:
Ugh, seeing `threading` in library code gives me the willies.  I know you've just moved it here from `_field.py`, bu has anybody done a measurement of the RAM cost over time of this cache?  It seems like a *huge* leak.

Reply:
While we're at it, why is it public?  Or even at module scope?

Reply:
This goes back to [3.8.1](https://github.com/zopefoundation/zope.schema/commit/6369d354c8d318bcda224d758243b6d31b598199). Previously it was setting a `_v_` attribute on each object as it was validated, but that lead to security issues, apparently.

As for a leak, the code is careful to push and pop object ids in a finally block, so the actual values shouldn't be leaking. If you're referring to the thread-local objects themselves leaking as threads are created and deleted, I don't think that's a problem. I believe this because I once [broke that for gevent and got complaints](https://github.com/gevent/gevent/issues/981). Under Python 2.7, local objects are stored in a Thread `__dict__` and so [automatically go away when the thread does](https://github.com/python/cpython/blob/2.7/Lib/_threading_local.py), and under 3.4 they are explicitly handled with [weakrefs](https://github.com/python/cpython/blob/3.4/Lib/_threading_local.py).

As for the scope, name, and visibility, this is used from a function, so it either needs to be module scoped or added as a hidden kwarg default to the function. I can do that, or just stick an `_` in front of it.

Also, since we're looking at it, I would like to change it to not use the object's `__dict__` directly. On PyPy, at least, those are optimized for `str` keys, and we're just shoving ints in there (PyPy may adjust that strategy, but I think it would have to do so anew for each thread.). Using an explicit attribute (a `set`) of the thread-local object would be more clear, IMO.

Reply:
> As for the scope, name, and visibility, this is used from a function, so it either needs to be module scoped or added as a hidden kwarg default to the function

I'm not sure why it can't just be a normal local variable declared inside the function.

Reply:
I had that thought, but then I realized it can be *implicitly* called again from some field that's an `Object` or subclass, or indirectly, such as from a `List` of `Object` fields. I went back and clarified my comment.

Reply:
Actually, given that we are popping the value off at the end, what is the purpose of the dang thing?  It isn't like `getSchemaValidationErrors` is recursing, or something.

Reply:
It's there to protect against cycles. 

If we don't have it, we get a test failure:
```
Error in test test_validate_w_cycles (zope.schema.tests.test__field.ObjectTests)
Traceback (most recent call last):
  File "///zope.schema36/bin/zope-testrunner", line 11, in <module>
    sys.exit(run())
  File "/////zope/testrunner/__init__.py", line 27, in run
    failed = run_internal(defaults, args, script_parts=script_parts, cwd=cwd)
  File "/////zope/testrunner/__init__.py", line 44, in run_internal
    runner.run()
  File "/////zope/testrunner/runner.py", line 160, in run
    self.run_tests()
  File "/////zope/testrunner/runner.py", line 242, in run_tests
    self.skipped, self.import_errors)
  File "/////zope/testrunner/runner.py", line 411, in run_layer
    import_errors)
  File "/////zope/testrunner/runner.py", line 334, in run_tests
    test(result)
  File "//lib/python3.6/unittest/case.py", line 653, in __call__
    return self.run(*args, **kwds)
  File "//lib/python3.6/unittest/case.py", line 613, in run
    self._feedErrorsToResult(result, outcome.errors)
  File "//lib/python3.6/unittest/case.py", line 543, in _feedErrorsToResult
    result.addError(test, exc_info)
  File "/////zope/testrunner/runner.py", line 788, in addError
    exc_info)
  File "/////zope/testrunner/formatter.py", line 335, in test_error
    self.print_traceback("Error in test %s" % test, exc_info)
  File "/////zope/testrunner/formatter.py", line 643, in print_traceback
    tb = self.format_traceback(exc_info)
  File "/////zope/testrunner/formatter.py", line 372, in format_traceback
    tb = "".join(traceback.format_exception(*exc_info))
  File "/////zope/testrunner/tb_format.py", line 57, in format_exception
    return fmt.formatException(t, v, tb)
  File "/////zope/exceptions/exceptionformatter.py", line 193, in formatException
    exc_line = self.formatExceptionOnly(etype, value)
  File "/////zope/exceptions/exceptionformatter.py", line 165, in formatExceptionOnly
    result = ''.join(traceback.format_exception_only(etype, value))
  File "//lib/python3.6/traceback.py", line 136, in format_exception_only
    return list(TracebackException(etype, value, None).format_exception_only())
  File "//lib/python3.6/traceback.py", line 486, in __init__
    _seen=_seen)
  File "//lib/python3.6/traceback.py", line 486, in __init__
    _seen=_seen)
  File "//lib/python3.6/traceback.py", line 486, in __init__
    _seen=_seen)
  [Previous line repeated 239 more times]
  File "//lib/python3.6/traceback.py", line 480, in __init__
    type(exc_value.__context__),
RecursionError: maximum recursion depth exceeded while calling a Python object
```

```python
    def test_validate_w_cycles(self):
        IUnit, Person, Unit = self._makeCycles()
        field = self._makeOne(schema=IUnit)
        person1 = Person(None)
        person2 = Person(None)
        unit = Unit(person1, [person1, person2])
        person1.unit = unit
        person2.unit = unit
        field.validate(unit)  # doesn't raise

    def _makeCycles(self):
        from zope.interface import Interface
        from zope.interface import implementer
        from zope.schema import Object
        from zope.schema import List
        from zope.schema._messageid import _

        class IUnit(Interface):
            """A schema that participate to a cycle"""
            boss = Object(
                schema=Interface,
                title=_("Boss"),
                description=_("Boss description"),
                required=False,
                )
            members = List(
                value_type=Object(schema=Interface),
                title=_("Member List"),
                description=_("Member list description"),
                required=False,
                )

        class IPerson(Interface):
            """A schema that participate to a cycle"""
            unit = Object(
                schema=IUnit,
                title=_("Unit"),
                description=_("Unit description"),
                required=False,
                )

        IUnit['boss'].schema = IPerson
        IUnit['members'].value_type.schema = IPerson

        @implementer(IUnit)
        class Unit(object):
            def __init__(self, person, person_list):
                self.boss = person
                self.members = person_list

        @implementer(IPerson)
        class Person(object):
            def __init__(self, unit):
                self.unit = unit

        return IUnit, Person, Unit
```

Reply:
I made it private using the kwarg-as-singleton-instance trick.